### DISCUSSION TITLE
mqtt_as startup behavior question

### DISCUSSION THREAD

**Discussion by nas2011** [micropython/micropython#11400] [General]
First, a huge thank you to Peter for a great library. I am connecting an ESP32 with a 433mhz receiver and transmitter to publish via MQTT so I can use home assistant for "smart" use of RF controlled ceiling fans. I started down the path of using the callback setup as I understand it fairly well, but tried out the event setup and it all worked. I have one quirk though that shows I don't understand enough what is going on.

When I lose connectivity, or when I restart the device, the "callback" I have running triggered by the message fires, even if no message has been received from a given topic. Why is that?

My use case is pretty much exactly the readme example, just trigger a different function. 

Code for clarity:

```py

config['ssid'] = myssid
config['wifi_pw'] = wifiPW
config['server'] = broker
config['user'] = mqttUser
config['password'] = mqttPW

async def messages(client):  # Respond to incoming messages
    async for topic, msg, retained in client.queue:
        print(topic,msg,retained)
        text = msg.decode()
        print(text)
        transmit(text) #transmit the corresponding rf code

async def up(client):  # Respond to connectivity being (re)established
    while True:
        await client.up.wait()  # Wait on an Event
        client.up.clear()
        await client.subscribe('home/fans', 1)  # renew subscriptions

async def main(client):
    await client.connect()
    for coroutine in (up, messages):
        asyncio.create_task(coroutine(client))
    while True:
        await asyncio.sleep(5)

config["queue_len"] = 1  # Use event interface with default queue size
MQTTClient.DEBUG = True  # Optional: print diagnostic messages
client = MQTTClient(config)

try:
    asyncio.run(main(client))
finally:
    client.close()
```

I'm not sure how it is that the messages function is being called when there is no message that was received. I literally start this up and yet it still fires. I'm admittedly very unclear on the whole event queue concept and I suspect it stems from that.

Also, I thought it was pretty awesome that with two libraries from Peter I was able to get this up in running in just a morning of work. Really great stuff. Thank you.

Thanks in advance for any guidance anyone is able to provide.

Best,
Nick

**Comment by beetlegigg** [comment]
I think its because you've used the retain message flag.  The broker always retains the last message that was sent from some client and will dish it out when a new client connects regardless of whether its been sent to the client on a previous connect.   I find retained messages can be a bit of a pain as a new connection will alway receive a retained message even if its an old and out of date message.  I got caught when I was lured into thinking an old retained message of a temperature reading was a current reading when I freshly connected to my broker and  thought the temperature sensor must have thrown a wobbly :-)

**Comment by nas2011** [comment]
Can you point out to me where I set the retain flag? 

I looked through the mqtt_as source and it looks like everywhere retain occurs the default is ```False``` and I looked for a way to set it but don't see one. I know I print the retain from the queue so I'm thinking I set it there, but again, quite confused with the whole async approach.

**Comment by nas2011** [comment]
Well this is embarrassing... I see that the retain is set on the publisher, not the subscriber. Looked a bit about mqtt an retain and see you can clear a retained message by sending a zero by message with retain being false. Did the trick.

Thanks!
